tools/slo-generator/slo_generator/report.py

```diff
@@ -0,0 +1,274 @@
+from slo_generator import utils
+
+LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(init=False)
+class SLOReport:
+    """SLO report dataclass. Compute an SLO report out of an SLO config and an
+    Error Budget Policy step.
+
+    Args:
+        config (dict): SLO configuration.
+        step (dict): Error budget policy step configuration.
+        timestamp (int): Timestamp.
+        client (obj): Existing backend client.
+        delete (bool): Backend delete action.
+    """
+    # pylint: disable=too-many-instance-attributes
+
+    # SLO
+    service_name: str
+    feature_name: str
+    slo_name: str
+    slo_target: float
+    slo_description: str
+    sli_measurement: float = 0
+    bad_events_count: int = 0
+    good_events_count: int = 0
+    gap: float
+
+    # Error budget
+    error_budget_policy_step_name: str
+    error_budget_target: float
+    error_budget_measurement: float
+    error_budget_burn_rate: float
+    error_budget_minutes: float
+    error_budget_remaining_minutes: float
+    error_minutes: float
+
+    # Error Budget step config
+    timestamp: int
+    timestamp_human: str
+    window: int
+    alert: bool
+    alerting_burn_rate_threshold: float
+    consequence_message: str
+
+    def __init__(self, config, step, timestamp, client=None, delete=False):
+
+        # Init dataclass fields from SLO config and Error Budget Policy
+        self.__set_fields(**config,
+                          **step,
+                          lambdas={
+                              'slo_target': float,
+                              'alerting_burn_rate_threshold': int
+                          })
+
+        # Set other fields
+        self.window = int(step['measurement_window_seconds'])
+        self.timestamp = int(timestamp)
+        self.timestamp_human = utils.get_human_time(timestamp)
+
+        # Get backend results
+        result = self.run_backend(config, client=client, delete=delete)
+        if result:
+            self.build(step, result)
+
+    def build(self, step, result):
+        """Compute all data necessary for the SLO report.
+
+        Args:
+            step (dict): Error Budget Policy step configuration.
+            result (obj): Backend result.
+            delete (bool, optional): Delete mode enabled.
+
+        See https://landing.google.com/sre/workbook/chapters/implementing-slos/
+        for details on the calculations.
+        """
+        info = self.__get_info()
+        LOGGER.debug(f"{info} | SLO report starting ...")
+
+        # SLI, Good count, Bad count, Gap from backend results
+        sli, good_count, bad_count = self.get_sli(result)
+        gap = sli - self.slo_target
+
+        # Error Budget calculations
+        eb_target = 1 - self.slo_target
+        eb_value = 1 - sli
+        eb_remaining_minutes = self.window * gap / 60
+        eb_target_minutes = self.window * eb_target / 60
+        eb_minutes = self.window * eb_value / 60
+        if eb_target == 0:
```

Comment:
Does this imply an error budget of 0? (Meaning any burn would be above budget.)

Reply:
No, this implies an error budget *burn rate* (`eb_burn_rate`) of 0, not the error budget value (which is `eb_value`) meaning we don't burn any error budget. 

This catches the case where the burn rate `eb_value / eb_target` cannot be computed because the denominator is 0.